File: three_sum.py
target = 0
# arr = [-1, 0, 1, 2, -1, -4]
# Expected Output: [[-1, -1, 2], [-1, 0, 1]]
# arr = [1, 2, 3, 4, 5]
# # Expected Output: []
# arr = [0, 0, 0, 0]
# # Expected Output: [[0, 0, 0]]
# arr = [-2, 0, 0, 2, 2]
# # Expected Output: [[-2, 0, 2]]
arr = [-5, 2, 3, 0, -2, 1, 4, -1]
# # Expected Output: [[-5, 1, 4], [-2, -1, 3], [-2, 0, 2], [-1, 0, 1]]


if len(arr) < 3:
    print(False)
else:
    arr.sort()
    res = []

    for i in range(len(arr) - 2):
        if i > 0 and arr[i] == arr[i - 1]:
            continue

        left = i + 1
        right = len(arr) - 1

        while left < right:
            total = arr[i] + arr[left] + arr[right]
            if total == target:
                res.append([arr[i], arr[left], arr[right]])
                left += 1
                right -= 1

                while left < right and arr[left] == arr[left - 1]:
                    left += 1
                while left < right and arr[right] == arr[right + 1]:
                    right -= 1

            elif total < target:
                left += 1
            else:
                right -= 1

    print(res)

File: three_sum.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        if (nums.length < 3)
            return res;

        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            int left = i + 1;
            int right = nums.length - 1;

            while (left < right) {
                int total = nums[i] + nums[left] + nums[right];

                if (total == 0) {
                    res.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    left++;
                    right--;

                    while (left < right && nums[left] == nums[left - 1]) left++;
                    while (left < right && nums[right] == nums[right + 1]) right--;
                } else if (total < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return res;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        int[][] testCases = {
            {-1, 0, 1, 2, -1, -4},
            {0, 0, 0, 0},
            {-2, 0, 0, 2, 2},
            {1, 2, 3, 4, 5},
            {-5, 2, 3, 0, -2, 1, 4, -1}
        };

        for (int[] testCase : testCases) {
            List<List<Integer>> result = solution.threeSum(testCase);
            System.out.println("Input: " + Arrays.toString(testCase));
            System.out.println("Triplets: " + result);
            System.out.println();
        }
    }
}